cornell-dti/course-plan
# Placeholder courses

### Summary 

This pull request adds placeholder courses to CoursePlan to eventually support templates.

- [x] Add placeholder courses that display when Firestore data added (see below)
- [x] Style appropriately (normal/compact view, hover states, etc.)
- [x] Add `FirestoreSemesterPlaceholder` type and make the semester course array have objects of either that type of `FireStoreSemesterCourse`
- [x] Ensure placeholders can be dragged and save changes when refreshed 
- [x] Add cautions to placeholders when starting semester does not match current semester, and handle both semester order cases

![image](https://user-images.githubusercontent.com/25535093/142068236-155c2eff-c6b0-430d-8789-87eae7e824f1.png)

Remaining TODOs:
- [ ] Click placeholder to open up a new course adding modal (future PR)

Remaining design questions
- [x] Does the compact view placeholder look ok?
- [x] Should hover state be on the entire card, or only the part right of the 6 dots (like it is now)?
- [x] What cursor should be used where? Pointer to look clickable, or hand to look draggable?
- [ ] How can the warning system be improved? Right now, it is hardcoded to check if the semester index matches where we think the placeholder should be, but this does not handle if users ever have differing numbers of semesters than what we expect (like winter and summer semesters, first year transfer, do not add all their old semesters, etc.).

### Firestore Schema 
Placeholders are stored in the list of courses in a given semester instead of being their own field, to ensure that they can be dragged and dropped anywhere on a user's plan (like between 2 normal courses. The schema of a placeholder generally matches that of the template data files (as seen in #564). The only fields that do anything right now are uniqueID, name (shows up on the placeholder), and startingSemester (determines if a warning shows up).
![image](https://user-images.githubusercontent.com/25535093/142042134-bad02351-4c1f-4aa4-86bc-66dbc493c46a.png)

### Test Plan 

1. Add a placeholder to a courses array in a semester following the above schema.
2. Confirm it can be dragged and dropped to other semesters while saving
3. Confirm a warning displays if the startingSemester prop does not match the semester the placeholder is in, on both "Newest" and "Oldest" toggles.
3. Confirm the design of the placeholder and warnings look good on both views, including hover states
4. Confirm that code changed to handle different types still works as expected, specifically...
- Credit counts on semesters are still accurate
- Placeholders are minimized when semesters are, and cleared when semesters are
- Selectable requirements work (and do not show placeholders in the dropdown)
- Colors for academic subjects can be changed
- Other warnings still work

### Notes 

Placeholders will only show up if data is manually added to the semesters Firestore collection, so this PR can be safely merged without a gatekeep
